### DISCUSSION TITLE
Writing an Interrupt Service Routine

### DISCUSSION THREAD

**Discussion by e76nystrom** [stnolting/neorv32#706] [Q&A] [answered]
I was looking at the disassembled code from the ISR of the SPI IRQ demo program and it does not appear to save the registers or use the mret.

Is this correct?

```00000188 <spi_irq_handler>:
     188:	ff010113          	addi	sp,sp,-16
     18c:	00112623          	sw	ra,12(sp)
     190:	80000537          	lui	a0,0x80000
     194:	00050513          	mv	a0,a0
     198:	150000ef          	jal	ra,2e8 <neorv32_spi_isr>
     19c:	00c12083          	lw	ra,12(sp)
     1a0:	01010113          	addi	sp,sp,16
     1a4:	00008067          	ret
```

I found that if I add the following declaration before the spi_irq_handler the registers will be saved and the routine exits with a mret.
```
static void spi_irq_handler(void) __attribute__ ((interrupt ("machine")));
```

**Comment by stnolting** [comment, marked as answer]
Hey @e76nystrom!

> I was looking at the disassembled code from the ISR of the SPI IRQ demo program and it does not appear to save the registers or use the mret.
Is this correct?

Yes, this is correct.

This "interrupt service routine" is not a "real" interrupt service routine. 😅 It is just a normal function that is called by the actual interrupt service routine, which is buried in the **NEORV32 Runtime Environment (RTE)**. All the register saving as well as the termination of the interrupt handler via `mret` is done by the RTE.

You can find more in-depth information about the RTE in the online data sheet: [:books: NEORV32 Runtime Environment](https://stnolting.github.io/neorv32/#_neorv32_runtime_environment)

**Reply by e76nystrom** [reply]
Thank you for the explanation.
